jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -0,0 +1,828 @@
+  const exprt &expr,
+  const jsont &json,
+  const optionalt<symbol_exprt> &given_length_expr,
+  const optionalt<std::string> &type_from_array,
+  det_creation_infot &info)
+{
+  PRECONDITION(has_array_type(expr, info.symbol_table));
+  PRECONDITION(can_cast_type<pointer_typet>(expr.type()));
+  const auto &element_type = static_cast<const typet &>(
+    to_pointer_type(expr.type()).subtype().find(ID_element_type));
+
+  exprt length_expr;
+  if(given_length_expr)
+    length_expr = *given_length_expr;
+  else
+  {
+    length_expr = info.allocate_objects.allocate_automatic_local_object(
+      java_int_type(), "det_array_length");
+    info.block.add(code_assignt{
+      length_expr, side_effect_expr_nondett{java_int_type(), info.loc}});
+    info.block.add(code_assumet{binary_predicate_exprt{
+      length_expr, ID_ge, from_integer(0, java_int_type())}});
+    allocate_array(expr, length_expr, info);
+  }
+  const json_arrayt json_array = get_untyped_array(json, element_type);
+  const auto number_of_elements =
+    from_integer(json_array.size(), java_int_type());
+  info.block.add(code_assumet{
+    binary_predicate_exprt{length_expr, ID_ge, number_of_elements}});
+  if(has_nondet_length(json))
+  {
+    info.block.add(code_assumet{binary_predicate_exprt{
+      length_expr,
+      ID_le,
+      from_integer(info.max_user_array_length, java_int_type())}});
+  }
+  else
+  {
+    info.block.add(code_assumet{
+      binary_predicate_exprt{length_expr, ID_le, number_of_elements}});
+  }
+  assign_array_data_component_from_json(expr, json, type_from_array, info);
+}
+
+/// One of the cases in the recursive algorithm: the case where \p expr
+/// represents a string.
+/// See \ref assign_from_json_rec.
+static void assign_string_from_json(
+  const jsont &json,
+  const exprt &expr,
+  det_creation_infot &info)
+{
+  const auto json_string = get_untyped_string(json);
+  PRECONDITION(json_string.is_string());
+  info.block.add(code_assignt{expr,
+                              get_or_create_string_literal_symbol(
+                                json_string.value, info.symbol_table, true)});
+}
+
+/// Helper function for \ref assign_struct_from_json which recursively assigns
+/// values to all of the fields of the Java object represented by \p expr (the
+/// components of its type and all of its parent types).
+static void assign_struct_components_from_json(
+  const exprt &expr,
+  const jsont &json,
+  det_creation_infot &info)
+{
+  const java_class_typet &java_class_type =
+    to_java_class_type(namespacet{info.symbol_table}.follow(expr.type()));
+  for(const auto &component : java_class_type.components())
+  {
+    const irep_idt &component_name = component.get_name();
+    if(
+      component_name == "@class_identifier" ||
+      component_name == "cproverMonitorCount")
+    {
+      continue;
+    }
+    member_exprt member_expr{expr, component_name, component.type()};
+    if(component_name[0] == '@') // component is parent struct type
```

Comment:
Ah, I should have made this clearer than just the short comment. If `component_name` starts with `@` then it has a struct type representing the parent class. I can't send this back to `assign_from_json_rec` because the latter expects pointer types or primitives, but not structs. (And I don't want to give it a struct case because I think it's confusing and requires passing around a boolean or defining more functions to keep this "struct storing components of the parent" as a special case - the object factory does it like this and it's not pretty.) All other components are fields that are declared in `java_class_type`, and those have pointer types and go back to the recursive case.

Reply:
To give an example, `MyClassWithInheritedFields` from the regression tests has the following components:
```
  * components: 
    0: 
      * type: struct_tag
          * identifier: java::com.diffblue.example.MyParent
      * name: @com.diffblue.example.MyParent
      * base_name: @com.diffblue.example.MyParent
      * pretty_name: @com.diffblue.example.MyParent
    1: 
      * type: signedbv
          * width: 32
      * name: childField
      * base_name: childField
      * access: default
      * pretty_name: childField
      * final: 0
```
Component 0 is the parent type, and component 1 is a pointer to `childField`.
We then call `assign_struct_components_from_json` on `MyParent`, which gives us the components
```
  * components: 
    0: 
      * type: struct_tag
          * identifier: java::java.lang.Object
      * name: @java.lang.Object
      * base_name: @java.lang.Object
      * pretty_name: @java.lang.Object
    1: 
      * type: signedbv
          * width: 32
      * name: parentField
      * base_name: parentField
      * access: default
      * pretty_name: parentField
      * final: 0
    2: 
      * type: pointer
          * width: 64
          * #reference: 1
          0: struct_tag
              * identifier: java::java.lang.String
              * #base_name: java.lang.String
      * name: parentString
      * base_name: parentString
      * access: default
      * pretty_name: parentString
      * final: 0
```
So we find all the fields declared in `MyParent` and also see that it inherits directly from Object. Then we call this function one more time on the `Object` type.